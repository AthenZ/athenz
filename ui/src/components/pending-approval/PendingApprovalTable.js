/*
 * Copyright 2020 Verizon Media
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import React, { Fragment } from 'react';
import styled from '@emotion/styled';
import { colors } from '../denali/styles';
import PendingApprovalTableRow from './PendingApprovalTableRow';
import PendingApprovalTableHeader from './PendingApprovalTableHeader';
import Color from '../../components/denali/Color';
import DateUtils from '../utils/DateUtils';
import RequestUtils from '../utils/RequestUtils';
const TableHeader = styled.th`
    border-bottom: 2px solid ${colors.grey500};
    color: ${colors.grey600};
    font-weight: 600;
    font-size: 0.7rem;
    padding-bottom: 5px;
    vertical-align: top;
    text-transform: uppercase;
    padding: 5px 0 5px 15px;
    text-align: left;
    white-space: nowrap;
`;

const TableHeaderDomain = styled.th`
    border-bottom: 2px solid ${colors.grey500};
    color: ${colors.grey600};
    font-weight: 600;
    font-size: 0.7rem;
    padding-bottom: 5px;
    vertical-align: top;
    text-transform: uppercase;
    padding: 5px 0 5px 0px;
    text-align: left;
`;

const ApproveTableHeader = styled.th`
    border-bottom: 2px solid ${colors.grey500};
    color: ${colors.grey600};
    font-weight: 600;
    font-size: 0.7rem;
    vertical-align: top;
    text-transform: uppercase;
    text-align: center;
    position: absolute;
    width: 15em;
    right: 13em;
    z-index: 1;
    height: 15px;
    padding-bottom: 5px;
    padding: 5px 0 5px 15px;
    white-space: nowrap;
`;

const RejectTableHeader = styled.th`
    border-bottom: 2px solid ${colors.grey500};
    color: ${colors.grey600};
    font-weight: 600;
    font-size: 0.7rem;
    vertical-align: top;
    text-transform: uppercase;
    text-align: center;
    border-right: none;
    position: absolute;
    width: 14em;
    right: 0em;
    z-index: 1;
    height: 15px;
    padding-bottom: 5px;
    padding: 5px 0 5px 15px;
    white-space: nowrap;
`;

const DomainListTable = styled.table`
    margin-top: 20px;
    width: 100%;
    border-spacing: 0;
`;

export default class PendingApprovalTable extends React.Component {
    constructor(props) {
        super(props);
        this.api = props.api;
        this.state = {
            checkedAll: false,
            pendingMap: this.props.pendingData,
            checkedList: [],
            selectAllAudit: '',
            selectAllDateExpiry: '',
            selectAllDateReviewReminder: '',
        };
        this.checkAllBoxOnchange = this.checkAllBoxOnchange.bind(this);
        this.dateUtils = new DateUtils();
    }

    auditRefChange(key, event) {
        if (key === 'SelectAll') {
            this.setState({
                selectAllAudit: event.target.value,
            });
        } else {
            const pendingMap = this.state.pendingMap;
            pendingMap[key].auditRef = event.target.value;
            this.setState({
                pendingMap: pendingMap,
            });
        }
    }

    dateChange(key, date, type) {
        if (date && date.length > 0) {
            date = this.dateUtils.uxDatetimeToRDLTimestamp(date);
            if (key === 'SelectAll') {
                if (type === 'expiry') {
                    this.setState({
                        selectAllDateExpiry: date,
                    });
                } else {
                    this.setState({
                        selectAllDateReviewReminder: date,
                    });
                }
            } else {
                const pendingMap = this.state.pendingMap;
                if (type === 'expiry') {
                    pendingMap[key].expiryDate = date;
                } else {
                    pendingMap[key].reviewReminder = date;
                }
                this.setState({
                    pendingMap: pendingMap,
                });
            }
        }
    }

    checkedChange(key) {
        const pos = this.state.checkedList.indexOf(key);
        let newList = this.state.checkedList;
        let pendingMap = this.state.pendingMap;
        if (pos === -1) {
            newList.push(key);
            pendingMap[key].auditRefMissing = false;
            if (newList.length === Object.keys(this.state.pendingMap).length) {
                this.setState({
                    checkedList: newList,
                    checkedAll: true,
                    pendingMap: pendingMap,
                });
            } else {
                this.setState({
                    checkedList: newList,
                    checkedAll: 2,
                    pendingMap: pendingMap,
                });
            }
        } else {
            newList.splice(pos, 1);

            if (newList.length === 0) {
                this.setState({
                    checkedList: newList,
                    checkedAll: false,
                    selectAllAuditMissing: false,
                });
            } else {
                this.setState({
                    checkedList: newList,
                    checkedAll: 2,
                });
            }
        }
    }

    initialLoad() {
        this.api
            .getPendingDomainMembersList()
            .then((data) => {
                this.setState({
                    pendingMap: data,
                    checkedList: [],
                    selectAllAuditMissing: false,
                    selectAllDateExpiry: '',
                    selectAllReviewReminder: '',
                });
            })
            .catch((err) => {
                this.setState({
                    errorMessage: RequestUtils.xhrErrorCheckHelper(err),
                });
            });
    }

    checkAllBoxOnchange() {
        let newList = [];
        if (this.state.checkedAll) {
            this.setState({
                checkedAll: false,
                checkedList: newList,
                selectAllAuditMissing: false,
            });
        } else {
            newList = Object.keys(this.state.pendingMap);
            this.setState({
                checkedList: newList,
                checkedAll: true,
            });
        }
    }

    pendingDecision(key, approved) {
        if (key === 'SelectAll') {
            if (this.state.selectAllAudit === '') {
                this.setState({
                    selectAllAuditMissing: true,
                });
            } else {
                let promises = [];
                this.state.checkedList.forEach((key) => {
                    let membership = {
                        memberName: this.state.pendingMap[key].memberName,
                        approved,
                        expiration: this.state.selectAllDateExpiry,
                        reviewReminder: this.state.selectAllDateReviewReminder,
                    };
                    promises.push(
                        this.api.processPending(
                            this.state.pendingMap[key].domainName,
                            this.state.pendingMap[key].roleName,
                            this.state.pendingMap[key].memberName,
                            this.state.selectAllAudit,
                            this.state.pendingMap[key].category,
                            membership,
                            this.props._csrf
                        )
                    );
                });
                Promise.all(promises)
                    .then(() => {
                        this.initialLoad();
                    })
                    .catch((err) => {
                        this.setState({
                            errorMessage: RequestUtils.xhrErrorCheckHelper(err),
                        });
                    });
            }
        } else {
            let auditRef = this.state.pendingMap[key].auditRef;
            if (auditRef === '') {
                let pendingMap = this.state.pendingMap;
                pendingMap[key].auditRefMissing = true;
                this.setState({
                    pendingMap: pendingMap,
                });
            } else {
                let membership = {
                    memberName: this.state.pendingMap[key].memberName,
                    approved,
                    expiration: this.state.pendingMap[key].expiryDate,
                    reviewReminder: this.state.pendingMap[key].reviewReminder,
                };
                this.api
                    .processPending(
                        this.state.pendingMap[key].domainName,
                        this.state.pendingMap[key].roleName,
                        this.state.pendingMap[key].memberName,
                        this.state.pendingMap[key].auditRef,
                        this.state.pendingMap[key].category,
                        membership,
                        this.props._csrf
                    )
                    .then(() => {
                        this.initialLoad();
                    })
                    .catch((err) => {
                        this.setState({
                            errorMessage: RequestUtils.xhrErrorCheckHelper(err),
                        });
                    });
            }
        }
    }

    render() {
        let contents = [];
        let pendingDecision = this.pendingDecision.bind(this);
        let auditRefChange = this.auditRefChange.bind(this);
        let dateChange = this.dateChange.bind(this);
        let checkedClick = this.checkedChange.bind(this);
        if (this.state.pendingMap) {
            let i = 0;
            Object.keys(this.state.pendingMap).forEach((keyInMap) => {
                let pending = this.state.pendingMap[keyInMap];
                const category = pending.category;
                const domainName = pending.domainName;
                const memberName = pending.memberName;
                const defaultExpiration = pending.expiryDate;
                const defaultReviewReminder = pending.reviewReminder;
                let color = '';
                if (i % 2 === 0) {
                    color = colors.row;
                }
                i += 1;
                const roleName = pending.roleName;
                const key = domainName + memberName + roleName;
                let checked;
                checked = this.state.checkedList.indexOf(key) !== -1;
                const userComment = pending.userComment;
                contents.push(
                    <PendingApprovalTableRow
                        category={category}
                        domainName={domainName}
                        memberName={memberName}
                        roleName={roleName}
                        userComment={userComment}
                        color={color}
                        checked={checked}
                        auditRefChange={auditRefChange}
                        dateChange={dateChange}
                        checkedClick={checkedClick}
                        pendingDecision={pendingDecision}
                        auditRefMissing={
                            this.state.pendingMap[key].auditRefMissing
                        }
                        key={domainName + roleName + memberName}
                        defaultExpiration={defaultExpiration}
                        defaultReviewReminder={defaultReviewReminder}
                        clearExpiry={this.state.pendingMap[key].expiryDate}
                        clearReviewReminder={
                            this.state.pendingMap[key].reviewReminder
                        }
                        requestPrincipal={
                            this.state.pendingMap[key].requestPrincipal
                        }
                        requestTime={this.state.pendingMap[key].requestTime}
                        requestPrincipalFull={
                            this.state.pendingMap[key].requestPrincipalFull
                        }
                        memberNameFull={
                            this.state.pendingMap[key].memberNameFull
                        }
                        requestedExpiry={this.state.pendingMap[key].expiryDate}
                        requestedReviewReminder={
                            this.state.pendingMap[key].reviewReminder
                        }
                    />
                );
            });
        }
        return (
            <Fragment>
                {this.state.errorMessage && (
                    <Color name={'red600'}>{this.state.errorMessage}</Color>
                )}
                <DomainListTable data-testid='pending-approval-table'>
                    <thead>
                        <PendingApprovalTableHeader
                            checked={this.state.checkedAll}
                            onChange={this.checkAllBoxOnchange}
                            checkedList={this.state.checkedList}
                            pendingDecision={pendingDecision}
                            auditRefChange={auditRefChange}
                            dateChange={dateChange}
                            auditRefMissing={this.state.selectAllAuditMissing}
                            api={this.api}
                            clearExpiry={this.state.selectAllDateExpiry}
                            clearReviewReminder={
                                this.state.selectAllDateReviewReminder
                            }
                        />
                        <tr>
                            <TableHeader />
                            <TableHeaderDomain>Domain</TableHeaderDomain>
                            <TableHeader>Type</TableHeader>
                            <TableHeader>Name</TableHeader>
                            <TableHeader>Member</TableHeader>
                            <TableHeader colSpan={2}>User Comment</TableHeader>
                            <TableHeader>Requester</TableHeader>
                            <TableHeader>Request Time</TableHeader>
                            <TableHeader>Audit Reference</TableHeader>
                            <TableHeader>Expiration Date</TableHeader>
                            <TableHeader>Reminder Date</TableHeader>
                            <ApproveTableHeader>Approve</ApproveTableHeader>
                            <RejectTableHeader>Reject</RejectTableHeader>
                        </tr>
                    </thead>
                    <tbody>{contents}</tbody>
                </DomainListTable>
            </Fragment>
        );
    }
}
