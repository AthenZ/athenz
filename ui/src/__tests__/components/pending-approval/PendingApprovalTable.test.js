/*
 * Copyright 2020 Verizon Media
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import React from 'react';
import { render } from '@testing-library/react';
import PendingApprovalTable from '../../../components/pending-approval/PendingApprovalTable';

describe('PendingApprovalTable', () => {
    it('should render', () => {
        const query = {
            domain: 'dom',
        };
        const domains = ['athens', 'athens.ci'];
        const userId = 'pgote';
        const domain = 'home.pgote';
        const api = {};
        const { getByTestId } = render(
            <PendingApprovalTable
                domains={domains}
                req='req'
                userId={userId}
                query={query}
                domain={domain}
                pendingData={undefined}
                domainResult={[]}
                api={api}
            />
        );

        const pendingapprovaltable = getByTestId('pending-approval-table');
        expect(pendingapprovaltable).toMatchSnapshot();
    });
});
