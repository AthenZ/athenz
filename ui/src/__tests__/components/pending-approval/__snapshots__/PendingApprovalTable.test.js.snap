// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PendingApprovalTable should render 1`] = `
.emotion-0 {
  margin-top: 20px;
  width: 100%;
  border-spacing: 0;
}

.emotion-2 {
  border-bottom: 2px solid #d5d5d5;
  color: #9a9a9a;
  font-weight: 600;
  padding-bottom: 5px;
  vertical-align: middle;
  text-transform: uppercase;
  text-align: center;
  border-right: none;
  padding: 5px 0 5px 15px;
  width: 1%;
}

.emotion-4 {
  -webkit-backface-visibility: hidden;
  backface-visibility: hidden;
  box-sizing: border-box;
  display: inline-block;
  height: 16px;
  line-height: 16px;
  margin: 3px;
  outline: 0;
  position: relative;
  vertical-align: middle;
}

.emotion-4>input[type='checkbox'] {
  box-sizing: border-box;
  cursor: pointer;
  display: inline-block;
  position: absolute;
  top: 0;
  visibility: hidden;
}

.emotion-4>input[type='checkbox']+label {
  box-sizing: border-box;
  cursor: pointer;
  font-family: Helvetica,Arial,sans-serif;
  font-size: 14px;
  font-weight: 300;
  height: 16px;
  line-height: 16px;
  padding: 0 10px 0 0;
  vertical-align: bottom;
  white-space: nowrap;
}

.emotion-4>input[type='checkbox']+label:before {
  background: #fff;
  border: 2px solid rgba(54,151,242,0.5);
  border-radius: 2px;
  box-sizing: border-box;
  content: '';
  cursor: pointer;
  display: inline-block;
  height: 16px;
  line-height: 16px;
  margin: 0px 3px -3px 0px;
  -webkit-transition: all 0.25s ease-in;
  transition: all 0.25s ease-in;
  width: 16px;
}

.emotion-4>input[type='checkbox']:hover+label:before {
  border: 2px solid #3697f2;
}

.emotion-4>input[type='checkbox']:focus+label:before {
  border: 2px solid rgba(54,151,242,0.5);
}

.emotion-4>input[type='checkbox'][data-partial='true']+label:before,
.emotion-4>input[type='checkbox']:checked+label:before {
  border: 2px solid #3697f2;
}

.emotion-4>input[type='checkbox']:disabled+label {
  color: #d5d5d5;
  cursor: not-allowed;
}

.emotion-4>input[type='checkbox']:disabled+label:before {
  border: 2px solid #d5d5d5;
}

.emotion-4>input[type='checkbox']:checked:not([data-partial='true'])+label:before {
  background: #3697f2;
}

.emotion-4>input[type='checkbox'][data-partial='true']+label:after {
  border: solid #3697f2;
  border-width: 2px 0 0 0;
  box-sizing: border-box;
  content: '';
  cursor: pointer;
  height: 16px;
  line-height: 16px;
  left: 4px;
  opacity: 1;
  position: absolute;
  top: 7px;
  width: 8px;
}

.emotion-4>input[type='checkbox']:checked:not([data-partial='true'])+label:after {
  border: solid #fff;
  border-width: 0 2px 2px 0;
  box-sizing: border-box;
  content: '';
  cursor: pointer;
  display: block;
  height: 9px;
  line-height: 16px;
  left: 6px;
  opacity: 1;
  position: absolute;
  top: 3px;
  -webkit-transform: rotate(45deg);
  -moz-transform: rotate(45deg);
  -ms-transform: rotate(45deg);
  transform: rotate(45deg);
  width: 4.5px;
}

.emotion-5 {
  border-bottom: 2px solid #d5d5d5;
  color: #9a9a9a;
  font-weight: 600;
  font-size: 0.7rem;
  padding-bottom: 5px;
  vertical-align: middle;
  text-transform: uppercase;
  padding: 5px 0 5px 0px;
  text-align: left;
}

.emotion-7 {
  border-bottom: 2px solid #d5d5d5;
  color: #9a9a9a;
  font-weight: 600;
  font-size: 0.7rem;
  padding-bottom: 5px;
  vertical-align: top;
  text-transform: uppercase;
  padding: 5px 0 5px 15px;
  text-align: left;
  white-space: nowrap;
}

.emotion-17 {
  border-bottom: 2px solid #d5d5d5;
  font-weight: 600;
  font-size: 0.7rem;
  padding-bottom: 5px;
  vertical-align: bottom;
  padding: 5px 0 5px 15px;
  text-align: left;
}

.emotion-19 {
  font-family: Helvetica,Arial,sans-serif;
  font-size: 14px;
  font-weight: 300;
  box-sizing: border-box;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-box-flex-flow: column nowrap;
  -webkit-flex-flow: column nowrap;
  -ms-flex-flow: column nowrap;
  flex-flow: column nowrap;
  position: relative;
  width: 100%;
}

.emotion-19>textarea {
  -webkit-appearance: none;
  -moz-appearance: none;
  -ms-appearance: none;
  appearance: none;
  background: rgba(53,112,244,0.05);
  border-top: 2px solid transparent;
  border-bottom: 2px solid transparent;
  border-right: transparent;
  border-left: transparent;
  border-radius: 2px;
  box-shadow: none;
  box-sizing: border-box;
  color: #303030;
  -webkit-flex: 1 0 auto;
  -ms-flex: 1 0 auto;
  flex: 1 0 auto;
  font: inherit;
  height: 37px;
  line-height: 1.5;
  margin: 0;
  outline: none;
  padding: 0.25em 0.75rem;
  text-align: left;
  -webkit-transition: background-color 0.2s ease-in-out,color 0.2s ease-in-out,border 0.2s ease-in-out;
  transition: background-color 0.2s ease-in-out,color 0.2s ease-in-out,border 0.2s ease-in-out;
  width: auto;
}

.emotion-19>textarea::-webkit-input-placeholder {
  color: rgba(48,48,48,0.6);
}

.emotion-19>textarea::-moz-placeholder {
  color: rgba(48,48,48,0.6);
}

.emotion-19>textarea:-ms-input-placeholder {
  color: rgba(48,48,48,0.6);
}

.emotion-19>textarea::placeholder {
  color: rgba(48,48,48,0.6);
}

.emotion-19>textarea:active,
.emotion-19>textarea:focus {
  background: rgba(53,112,244,0.05);
  border-bottom: 2px solid #3570f4;
  color: #303030;
}

.emotion-19>textarea:invalid {
  background: rgba(53,112,244,0.05);
  border-bottom: 2px solid #d01111;
  color: #303030;
}

.emotion-19>textarea[disabled] {
  background: rgba(48,48,48,0.05);
  border-bottom: 2px solid rgba(48,48,48,0.05);
  color: rgba(48,48,48,0.25);
  cursor: not-allowed;
}

.emotion-19 .message {
  font-size: 12px;
  color: #303030;
  left: 0;
  line-height: 1.8;
  position: absolute;
  top: 2.571rem;
}

.emotion-20>textarea {
  resize: vertical;
}

.emotion-24>div input {
  position: relative;
  font: 300 14px HelveticaNeue-Reg,Helvetica,Arial,sans-serif;
  background-color: #d5d5d5;
  box-shadow: none;
  color: rgb(48, 48, 48);
  height: 16px;
  min-width: 50px;
  text-align: left;
  border-width: 2px;
  border-style: solid;
  border-color: transparent;
  border-image: initial;
  border-radius: 2px;
  -webkit-flex: 1 0 auto;
  -ms-flex: 1 0 auto;
  flex: 1 0 auto;
  margin: 0px;
  margin-top: 5px;
  outline: none;
  padding: 0.6em 12px;
  -webkit-transition: background-color 0.2s ease-in-out 0s,color 0.2s ease-in-out 0s,border 0.2s ease-in-out 0s;
  transition: background-color 0.2s ease-in-out 0s,color 0.2s ease-in-out 0s,border 0.2s ease-in-out 0s;
  width: 10em;
}

.emotion-26 {
  fill: #188fff;
  cursor: pointer;
  vertical-align: text-bottom;
}

.emotion-32 {
  border-bottom: 2px solid #d5d5d5;
  color: #9a9a9a;
  font-weight: 600;
  vertical-align: middle;
  text-transform: uppercase;
  text-align: center;
  position: absolute;
  width: 11em;
  right: 11em;
  z-index: 1;
  height: 42px;
  background-color: rgba(255, 255, 255, 1);
  padding-bottom: 5px;
  padding: 5px 0 5px 15px;
}

.emotion-34 {
  margin-top: 10px;
}

.emotion-36 {
  fill: #d5d5d5;
  cursor: pointer;
  vertical-align: text-bottom;
}

.emotion-37 {
  border-bottom: 2px solid #d5d5d5;
  color: #9a9a9a;
  font-weight: 600;
  vertical-align: middle;
  text-transform: uppercase;
  text-align: center;
  border-right: none;
  position: absolute;
  width: 11em;
  right: 0em;
  z-index: 1;
  height: 42px;
  background-color: rgba(255, 255, 255, 1);
  padding-bottom: 5px;
  padding: 5px 0 5px 15px;
}

.emotion-44 {
  border-bottom: 2px solid #d5d5d5;
  color: #9a9a9a;
  font-weight: 600;
  font-size: 0.7rem;
  padding-bottom: 5px;
  vertical-align: top;
  text-transform: uppercase;
  padding: 5px 0 5px 0px;
  text-align: left;
}

.emotion-64 {
  border-bottom: 2px solid #d5d5d5;
  color: #9a9a9a;
  font-weight: 600;
  font-size: 0.7rem;
  vertical-align: top;
  text-transform: uppercase;
  text-align: center;
  position: absolute;
  width: 15em;
  right: 13em;
  z-index: 1;
  height: 15px;
  padding-bottom: 5px;
  padding: 5px 0 5px 15px;
  white-space: nowrap;
}

.emotion-66 {
  border-bottom: 2px solid #d5d5d5;
  color: #9a9a9a;
  font-weight: 600;
  font-size: 0.7rem;
  vertical-align: top;
  text-transform: uppercase;
  text-align: center;
  border-right: none;
  position: absolute;
  width: 14em;
  right: 0em;
  z-index: 1;
  height: 15px;
  padding-bottom: 5px;
  padding: 5px 0 5px 15px;
  white-space: nowrap;
}

<table
  class="emotion-0 emotion-1"
  data-testid="pending-approval-table"
>
  <thead>
    <tr
      data-testid="pending-approval-table-header"
    >
      <th
        class="emotion-2 emotion-3"
      >
        <div
          class="emotion-4 denali-checkbox"
          data-testid="checkbox-wrapper"
        >
          <input
            data-partial="false"
            id="select-all-header"
            name="checkbox-header"
            type="checkbox"
          />
          <label
            for="select-all-header"
          />
        </div>
      </th>
      <th
        class="emotion-5 emotion-6"
      >
        <div>
           Select All 
        </div>
      </th>
      <th
        class="emotion-7 emotion-8"
      />
      <th
        class="emotion-7 emotion-8"
      />
      <th
        class="emotion-7 emotion-8"
        colspan="3"
      />
      <th
        class="emotion-7 emotion-8"
      />
      <th
        class="emotion-7 emotion-8"
      />
      <th
        class="emotion-17 emotion-18"
      >
        <div
          class="emotion-19 denali-textarea emotion-20 emotion-21"
          data-testid="textarea-wrapper"
        >
          <textarea
            data-testid="textarea"
            disabled=""
            placeholder="Justification"
            rows="1"
          />
        </div>
      </th>
      <th
        class="emotion-17 emotion-18"
      >
        <div
          class="emotion-24 emotion-25"
          disabled=""
        >
          <div
            class="fp-workflowHeaderExpiry"
          >
            <input
              class="flatpickr-input"
              data-input=""
              data-testid="flatPicker"
              placeholder="Expiration (Optional)"
              type="hidden"
              value=""
            />
            <input
              class=" form-control input"
              placeholder="Expiration (Optional)"
              readonly="readonly"
              tabindex="0"
              type="text"
            />
            <a
              data-clear=""
            >
              <svg
                class="emotion-26"
                data-testid="icon"
                height="1.25em"
                viewBox="0 0 1024 1024"
                width="1.25em"
              >
                <title>
                  close
                </title>
                <path
                  d="M571.733 512l226.133-224c7.901-8.071 12.777-19.131 12.777-31.33 0-9.87-3.192-18.994-8.599-26.397l0.088 0.127c-7.382-10.88-19.691-17.934-33.649-17.934-12.060 0-22.89 5.267-30.315 13.626l-0.036 0.042-226.133 226.133-226.133-226.133c-7.461-8.401-18.291-13.668-30.351-13.668-13.957 0-26.267 7.055-33.558 17.793l-0.091 0.142c-5.319 7.276-8.51 16.4-8.51 26.27 0 12.199 4.876 23.259 12.785 31.338l-0.008-0.008 226.133 224-226.133 224c-7.901 8.071-12.777 19.131-12.777 31.33 0 9.87 3.192 18.994 8.599 26.397l-0.088-0.127c7.382 10.88 19.691 17.934 33.649 17.934 12.060 0 22.89-5.267 30.315-13.626l0.036-0.042 226.133-226.133 226.133 226.133c7.461 8.401 18.291 13.668 30.351 13.668 13.957 0 26.267-7.055 33.558-17.793l0.091-0.142c5.319-7.276 8.51-16.4 8.51-26.27 0-12.199-4.876-23.259-12.785-31.338l0.008 0.008z"
                />
              </svg>
            </a>
          </div>
        </div>
      </th>
      <th
        class="emotion-17 emotion-18"
      >
        <div
          class="emotion-24 emotion-25"
          disabled=""
        >
          <div
            class="fp-workflowHeaderReviewReminder"
          >
            <input
              class="flatpickr-input"
              data-input=""
              data-testid="flatPicker"
              placeholder="Reminder (Optional)"
              type="hidden"
              value=""
            />
            <input
              class=" form-control input"
              placeholder="Reminder (Optional)"
              readonly="readonly"
              tabindex="0"
              type="text"
            />
            <a
              data-clear=""
            >
              <svg
                class="emotion-26"
                data-testid="icon"
                height="1.25em"
                viewBox="0 0 1024 1024"
                width="1.25em"
              >
                <title>
                  close
                </title>
                <path
                  d="M571.733 512l226.133-224c7.901-8.071 12.777-19.131 12.777-31.33 0-9.87-3.192-18.994-8.599-26.397l0.088 0.127c-7.382-10.88-19.691-17.934-33.649-17.934-12.060 0-22.89 5.267-30.315 13.626l-0.036 0.042-226.133 226.133-226.133-226.133c-7.461-8.401-18.291-13.668-30.351-13.668-13.957 0-26.267 7.055-33.558 17.793l-0.091 0.142c-5.319 7.276-8.51 16.4-8.51 26.27 0 12.199 4.876 23.259 12.785 31.338l-0.008-0.008 226.133 224-226.133 224c-7.901 8.071-12.777 19.131-12.777 31.33 0 9.87 3.192 18.994 8.599 26.397l-0.088-0.127c7.382 10.88 19.691 17.934 33.649 17.934 12.060 0 22.89-5.267 30.315-13.626l0.036-0.042 226.133-226.133 226.133 226.133c7.461 8.401 18.291 13.668 30.351 13.668 13.957 0 26.267-7.055 33.558-17.793l0.091-0.142c5.319-7.276 8.51-16.4 8.51-26.27 0-12.199-4.876-23.259-12.785-31.338l0.008 0.008z"
                />
              </svg>
            </a>
          </div>
        </div>
      </th>
      <th
        class="emotion-32 emotion-33"
      >
        <div
          class="emotion-34 emotion-35"
        >
          <svg
            class="emotion-36"
            data-testid="icon"
            height="1.25em"
            viewBox="0 0 1024 1024"
            width="1.25em"
          >
            <title>
              check-circle
            </title>
            <path
              d="M512 128c212.077 0 384 171.923 384 384s-171.923 384-384 384c-212.077 0-384-171.923-384-384v0c0-212.077 171.923-384 384-384v0zM512 42.667c-259.206 0-469.333 210.128-469.333 469.333s210.128 469.333 469.333 469.333c259.206 0 469.333-210.128 469.333-469.333v0c0-259.206-210.128-469.333-469.333-469.333v0z"
            />
            <path
              d="M396.16 691.413l-126.293-126.293c-8.506-8.066-13.798-19.448-13.798-32.066 0-9.118 2.763-17.59 7.498-24.625l-0.1 0.158c7.83-10.919 20.483-17.949 34.778-17.949 11.841 0 22.556 4.824 30.286 12.613l0.003 0.003 98.133 97.707 268.373-268.373c7.549-6.884 17.635-11.1 28.706-11.1 23.564 0 42.667 19.103 42.667 42.667 0 11.017-4.176 21.060-11.031 28.63l0.032-0.036-298.667 298.667c-7.733 7.795-18.45 12.621-30.293 12.621s-22.56-4.826-30.291-12.618l-0.003-0.003z"
            />
          </svg>
        </div>
      </th>
      <th
        class="emotion-37 emotion-38"
      >
        <div
          class="emotion-34 emotion-35"
        >
          <svg
            class="emotion-36"
            data-testid="icon"
            height="1.25em"
            viewBox="0 0 1024 1024"
            width="1.25em"
          >
            <title>
              decline
            </title>
            <path
              d="M512 42.667c-259.206 0-469.333 210.128-469.333 469.333s210.128 469.333 469.333 469.333c259.206 0 469.333-210.128 469.333-469.333v0c0.003-0.635 0.005-1.387 0.005-2.138 0-258.027-209.173-467.2-467.2-467.2-0.752 0-1.504 0.002-2.255 0.005l0.117-0zM130.133 467.2c22.95-192.169 184.962-339.729 381.44-339.729 90.864 0 174.356 31.559 240.114 84.314l-0.754-0.585-539.733 539.733c-52.076-65.227-83.563-148.885-83.563-239.896 0-15.446 0.907-30.681 2.671-45.653l-0.174 1.816zM554.667 893.867c-12.252 1.372-26.457 2.154-40.844 2.154-92.264 0-177.012-32.182-243.63-85.933l0.741 0.579 539.733-537.6c53.445 64.395 85.876 147.901 85.876 238.983 0 15.772-0.972 31.316-2.861 46.577l0.184-1.826c-22.136 177.253-161.672 316.060-337.376 336.891l-1.824 0.176z"
            />
          </svg>
        </div>
      </th>
    </tr>
    <tr>
      <th
        class="emotion-7 emotion-43"
      />
      <th
        class="emotion-44 emotion-45"
      >
        Domain
      </th>
      <th
        class="emotion-7 emotion-43"
      >
        Type
      </th>
      <th
        class="emotion-7 emotion-43"
      >
        Name
      </th>
      <th
        class="emotion-7 emotion-43"
      >
        Member
      </th>
      <th
        class="emotion-7 emotion-43"
        colspan="2"
      >
        User Comment
      </th>
      <th
        class="emotion-7 emotion-43"
      >
        Requester
      </th>
      <th
        class="emotion-7 emotion-43"
      >
        Request Time
      </th>
      <th
        class="emotion-7 emotion-43"
      >
        Audit Reference
      </th>
      <th
        class="emotion-7 emotion-43"
      >
        Expiration Date
      </th>
      <th
        class="emotion-7 emotion-43"
      >
        Reminder Date
      </th>
      <th
        class="emotion-64 emotion-65"
      >
        Approve
      </th>
      <th
        class="emotion-66 emotion-67"
      >
        Reject
      </th>
    </tr>
  </thead>
  <tbody />
</table>
`;
